### ISSUE TITLE
How are SPOS and FairNAS experiments actually done?

### ISSUE THREAD

**Issue by serser** [kcyu2014/eval-nas#4]
Hello, thanks for the impressive evaluation work! 

However, I am confused about the search space used for the experiments (in the supplementary of your ICLR paper) on one-shot NAS algorithms. As both handle MBV2-like block-wise search space, it requires some adjustment to apply them on cell-based NAS-Bench-101. 

According to `README.md` at the root of the repository, it says `bash scripts/nasbench-oneshot-search.sh search`  is used for one-shot evaluation. It seems to use the full NAS-Bench-101 search space. I cannot run into the very detail of the implementation, could you show how is it exactly done for single-path sampling? Just explain it in a nutshell?

Plus, in `search_policies/cnn/search_space/README.MD` a linearized search space is proposed for these methods, but it is nowhere used in any parts of code.

I am grateful for any help you can provide.

**Comment by serser** [comment]
Hello. I've run the code.  If I am not making mistake, the SPOS and FairNAS comparison in the paper shall be in the `NASbenchSearchSpace` and `fairnas_train_model_v1` is used for FairNAS training.

I again see that in `cnn_general_search_policies.py`, it says

```
Extend darts training method with FairNas strategy. It is not possible to use directly the FairNAS,
            but we can extend it into 2 method.
```

Could you explain what are exactly **two methods**? Are you referring to uniform sampling (topology sampling) and fairness sampling (operator sampling)?

As per my understanding and according to `fairnas_train_model_v1` in `fairnas_procedure.py`, the FariNAS execution shall be as follows,

- for each batch, sample a topology from the search space (a DAG cell: 7x7 adjacent matrix denoting edges, each vertex has three possible ops 1x1, 3x3 and maxpool, 2 out of 7 verticies are input and output), the result topology can have less vertices (say 6).
- `architect` is never used since it is initialized as `None`.
- Get model from `op_sampler` (in `nasbench_weight_sharing_policy.py`) which samples ops for the current topology. Three will be generated (there are 3 ops).
- Train each of 3 models one step with the same input, accumulate the weights
-  backpropagate to update the whole model

```python
    def op_sampler(self, model, architect, args):
        """
        Sample operation from model, used mainly for FairNAS procedure.
        :param model:
        :param architect:
        :param args:
        :return:
        """
        spec = self.model_spec
        ops = spec.ops
        avail_ops = self.search_space.available_ops
        try:
            op_vs_choice = np.tile(np.arange(len(avail_ops)), (len(ops)-2, 1))
            op_vs_choice = np.apply_along_axis(np.random.permutation, 1, op_vs_choice).transpose()
            for i in range(len(avail_ops)):
                new_ops = [avail_ops[ind] for ind in op_vs_choice[i]]
                spec.ops = ['input',] + new_ops + ['output']
                yield change_model_spec(model, spec)
        except ValueError as e:
            logging.warning(f'Op sampler: received exception {e}, return the original model without any op sampling.')
            yield model
```

To illustrate, I profiled one example step is like,

```
(Pdb) print(len(avail_ops))
3
(Pdb) print(len(ops))
6
(Pdb) print(op_vs_choice)
[[2 0 2 1]
 [0 2 0 2]
 [1 1 1 0]]
(Pdb) n
> /home/hadoop-automl/yu-eval-nas/search_policies/cnn/random_policy/nasbench_weight_sharing_policy.py(108)op_sampler()
-> new_ops = [avail_ops[ind] for ind in op_vs_choice[i]]
(Pdb) n
> /home/hadoop-automl/yu-eval-nas/search_policies/cnn/random_policy/nasbench_weight_sharing_policy.py(109)op_sampler()
-> spec.ops = ['input',] + new_ops + ['output']
(Pdb) print(new_ops)
['conv3x3-bn-relu', 'maxpool3x3', 'conv3x3-bn-relu', 'maxpool3x3']
(Pdb) print(spec)
Adjacent matrix: [[0, 1, 1, 0, 1, 1], [0, 0, 0, 1, 1, 0], [0, 0, 0, 1, 0, 0], [0, 0, 0, 0, 1, 0], [0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0]] 
Ops: ['input', 'maxpool3x3', 'conv3x3-bn-relu', 'maxpool3x3', 'conv1x1-bn-relu', 'output']
```

This literally says, a 6-vertex topology (2 of 6 are input and output) is sampled and 3 permutations (shown by 4x3 matrix `op_vs_choice`) out of 3^4 are used for a non-strict fairness training. This may explain the disparity of Kendall Tau calculation vs. what is reported in FairNAS. I would see a more strict but less feasible solution to loop all 3^4=81 possible permutations within the current topology. Please correct me if I am wrong.

P.S. I see that the new `NasBenchSearchSpaceFairNasTopology` if for the policy`topo_fairnas`.  Could you help me out with what is this for?

**Comment by kcyu2014** [comment]
Hi serser,

Thanks for your interest! Yes, we did adapt some one-shot NAS algorithms for them to work on NASBench 101.

For the single-path one-shot type approach, please check [this file](https://github.com/kcyu2014/eval-nas/blob/385376a3ef96336b54ee7e696af1d02b97aa5c32/search_policies/cnn/random_policy/nasbench_weight_sharing_policy.py#L39) for the logic. The `random_sampler` function implements this, essentially, random sample one architecture from the search space and train it. 

As for the linearized search space, that's not relevant to the ICLR paper, it's an experiment I tried to let NASBench-101 mimic mobile-net like search space. 

Let me know if you would like to know more.

P.S., I saw an email with more comments, but cannot find the reply on GitHub issue page yet. Is it because you delete the message? 

Best,
Kaicheng

**Comment by kcyu2014** [comment]
To your other question, FairNAS on NASBench-101 is indeed a quick prototype. Your understanding is exact, for each topology sampled, I mutate the operation as the original FairNAS policy. This is also the most direct application of FairNAS paper. Additional treatment should be considered as a FairNAS-v2 that improves the original paper. 

By the way, your proposal sounds interesting but I have some other comments. FairNAS points that you should train the weights in a fair way, that's why they always sample all the weights at one iteration. In NASBench-101, different topology does not infect the weights because unlike DARTS space where weights lie on the edge, the weights of NB101 lies on the node. 
That says, if at one iteration, we sample a full topology (no matter what's the edge connection), all the weights will be trained by the current operation sampler.

**Comment by serser** [comment]
@kcyu2014 Thanks for the detailed reply. I deleted the second question by mistake but I don't know how to find it back. I see that once the topology is chosen within the current batch, the operation sampler can be done in a fair way. But since the topology is sampled every batch, do really the weights easily transfer to the new topology? 

I've run SPOS vs. FairNAS with 50 epochs, I see that both supernets are not trained properly. This would make supernets fail to evaluate submodels. Just to double-check if I am doing all right, could you still find the log referring to supernet validation accuracy as well as sampled models‘ prediction vs. NB101 ground-truth?

**Comment by kcyu2014** [comment]
50 epochs are too short. In my next work (how to train your super net), I had a full comparison of fairnas v.s. Spos. It’s slightly better even in the current formulation. 

Cheers,
Kaicheng

**Comment by serser** [comment]
Thanks. But my question still remains unanswered, and could you please give me the searching logs?

**Comment by kcyu2014** [comment]
According to the other work, super-net accuracy (your question) does not correlate with the prediction power. Please check this paper (https://arxiv.org/pdf/2003.04276.pdf) page 8 Figure 9. 

Regarding the log for the super-net validation v.s. NB ground-truth I cannot find them since they are stored on a old server. 

Besides, to your first question, I do not see why topology sampling will make the weights transfer hard. This is exactly following randomNAS (or single-path one shot) approach on NB101.
